src/wallet/wallet.h

```diff
@@ -144,6 +144,81 @@ struct COutputEntry
     int vout;
 };
 
+/** An note outpoint */
+class JSOutPoint
+{
+public:
+    // Transaction hash
+    uint256 hash;
+    // Index into CTransaction.vjoinsplit
+    uint8_t js;
+    // Index into JSDescription fields of length ZC_NUM_JS_OUTPUTS
+    uint8_t n;
+
+    JSOutPoint() { SetNull(); }
+    JSOutPoint(uint256 h, uint8_t js, uint8_t n) : hash {h}, js {js}, n {n} { }
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        READWRITE(hash);
+        READWRITE(js);
+        READWRITE(n);
+    }
+
+    void SetNull() { hash.SetNull(); js = (uint8_t) -1; n = (uint8_t) -1; }
+    bool IsNull() const { return (hash.IsNull() && js == (uint8_t) -1 && n == (uint8_t) -1); }
```

Comment:
I actually decided to remove the `-1` checks entirely, because I didn't want to introduce that as an assumption (because these are actually unsigned ints). For the purposes I'm using it for, checking `hash.IsNull()` should be sufficient.